### ISSUE TITLE
yychar() for collecting characters

### ISSUE THREAD

**Issue by osa1** [LeoTestard/rustlex#51]
Being able to get characters matched in `.` pattern would make lexing strings much faster and easier. Say I have a string literal syntax that works like this: A literal can start with some number of `"`s and it ends with the same number of `"`s. Currently we can do something like:

```
        '"'+ => |l : &mut Lexer<R>| {
            l.delim_size = l.yystr().len();
            l.STRING();
            None
        }
```

Now, in `STRING` we have to match `.`s when `.` is not `"`, but the code is very slow because it allocates a new `String` just to add a character to the string buffer. E.g.

```
    STRING {
        . => |l : &mut Lexer<R>| {
            let str = l.yystr();
            l.string_buf.push_str(&str);
            None
        }
        '"'+ => ...
```

**Comment by LeoTestard** [comment]
> Now, in STRING we have to match .s when . is not ", but the code is very slow because it allocates a new String just to add a character to the string buffer. E.g.

This is correct, and this is indeed a problem. However, I'm just very insatisfied overall with the current way matched text is extracted and I'd rather rework it entirely rather than adding special cases. Something I have wanted to do for a long time is the ability to bind only sub-patterns, discarding the rest. Such a system could infer that the type of a variable that binds to the text matched by `.` is `char` instead of `String`… But the submatch capture itself is especially hard with deterministic automata...
